src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -1965,6 +2009,134 @@ PeerImp::onMessage (std::shared_ptr <protocol::TMHaveTransactionSet> const& m)
     }
 }
 
+void
+PeerImp::onMessage (std::shared_ptr <protocol::TMValidatorList> const& m)
+{
+    try
+    {
+        assert(supportsProtocolVersion("XRPL/2.1"));
+        auto const& manifest = m->manifest();
+        auto const& blob = m->blob();
+        auto const& signature = m->signature();
+        auto const version = m->version();
+        auto const hash = sha512Half(manifest, blob, signature, version);
+
+        JLOG(p_journal_.debug()) << "Received validator list from " <<
+            remote_address_.to_string() << " (" << id_ << ")";
+
+        if (! app_.getHashRouter ().addSuppressionPeer(hash, id_))
+        {
+            JLOG(p_journal_.debug()) <<
+                "ValidatorList: received duplicate validator list";
+            // Charging this fee here won't hurt the peer in the normal
+            // course of operation (ie. refresh every 5 minutes), but
+            // will add up if the peer is misbehaving.
+            fee_ = Resource::feeUnwantedData;
+            return;
+        }
+
+        auto const applyResult = app_.validators().applyListAndBroadcast (
+            manifest,
+            blob,
+            signature,
+            version,
+            remote_address_.to_string(),
+            hash,
+            app_.overlay(),
+            app_.getHashRouter());
+        auto const disp = applyResult.disposition;
+
+        JLOG(p_journal_.debug()) << "Processed validator list from " <<
+            (applyResult.publisherKey ? strHex(*applyResult.publisherKey) :
+            "unknown or invalid publisher") << " from " <<
+            remote_address_.to_string() << " (" << id_ << ") with result " <<
+            to_string(disp);
+
+        switch (disp)
+        {
+        case ListDisposition::accepted:
+            JLOG (p_journal_.debug()) <<
+                "Applied new validator list from peer " << remote_address_;
+            {
+                // if the remote node uses any of the same lists
+                // as this node, treat it as sane.
+                auto expected = Sanity::unknown;
+                sanity_.compare_exchange_strong(expected, Sanity::sane);
```

Comment:
I'm not sure I agree that we should treat such peers as sane. Sanity implies _more_.

Reply:
That's fair.